Utilities+LibCore: Port to LibMain + add syscall wrappers

kennethmyhra opened the pull request:
This ports the following utilities to LibMain:
- cal
- checksum
- chgrp
- chown
- chres
- cksum
- clear
- comm
- config
- cpp-lexer
- cpp-parser
- cpp-processor
- cut
- date
- dd

This also adds the following syscall wrappers:
- chown()
- getpwnam()
- getgrnam()
- clock_settime()

awesomekling commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -221,4 +221,12 @@ ErrorOr<void> chmod(StringView pathname, mode_t mode)
     return {};
 }
 
+ErrorOr<void> stat(StringView path, struct stat* statbuf)
```
We already have this syscall, with a slightly more convenient API :^)

```cpp
ErrorOr<struct stat> stat(StringView path);
```

kennethmyhra replied:
Nice, updated :)

kennethmyhra commented:
The PR is getting a bit large. I think it is probably better to continue porting another batch of utilities in a new PR

bgianfo commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -205,4 +205,27 @@ ErrorOr<String> gethostname()
 
+ErrorOr<void> chown(StringView pathname, uid_t uid, gid_t gid)
+{
+    int rc = ::chown(pathname.characters_without_null_termination(), uid, gid);
```
This is a bug waiting to happen... You have to ensure that these are null terminated, see implementations above that use the raw syscall so they can pass the explicit length instead.

Same goes with `chmod` below.

kennethmyhra replied:
Fixed by using raw syscall for `chown()` and `chmod()`.

bgianfo commented on Userland/Utilities/chmod.cpp:
```diff
@@ -67,14 +66,15 @@ int main(int argc, char** argv)
-        const char* access_string = argv[1];
+        const char* access_string = arguments.strings[1].characters_without_null_termination();
 
         while (*access_string != '\0') {
```
We should loop on the length of the StringView instead of assuming it t will be null terminated (it will be... But it's a bug to assume that with the StringView API)

kennethmyhra replied:
I was trying to avoid changing too much of the implementation logic since the bugs are probably going to hunt me into next year.

Would something like this be acceptable?
```cpp
const char* access_string = arguments.strings[1].to_string().characters();
```

kennethmyhra replied:
Struggled a bit with this. Ended up with the code below since the while loop would not recognize the contents with the one I proposed above.
```cpp
String temp = arguments.strings[1].to_string();
const char* access_string = temp.characters();
```

bgianfo commented on Userland/Utilities/chown.cpp:
```diff
@@ -55,7 +54,7 @@ int main(int argc, char** argv)
             new_gid = number.value();
         } else {
-            auto* group = getgrnam(parts[1].characters());
+            auto* group = getgrnam(parts[1].characters_without_null_termination());
```
Needs to be null terminated, unless you can implement a `getgrnam` that takes a StringView

kennethmyhra replied:
Updated like this, same for `getpwnam()` above:
```
auto* group = getgrnam(parts[1].to_string().characters());
```

kennethmyhra replied:
Added sycall wrappers for getpwnam and getgrnam.

bgianfo commented on Userland/Utilities/date.cpp:
```diff
@@ -30,21 +27,20 @@ int main(int argc, char** argv)
     if (set_date != nullptr) {
         auto number = String(set_date).to_uint();
 
+        outln("time from command line: {}", number.value());
```
Left over debugging code?

kennethmyhra replied:
Yes, removed

bgianfo commented on Userland/Utilities/dd.cpp:
```diff
@@ -142,36 +143,38 @@ int main(int argc, char** argv)
-        } else if (!strncmp(argv[a], "status=", 7)) {
-            if (handle_status_arguments(status, argv[a]) < 0) {
+        } else if (argument.starts_with("status=")) {
+            if (handle_status_arguments(status, argument.characters_without_null_termination()) < 0) {
```
Can we fix all of these handle functions to just take StringView's instead?

kennethmyhra replied:
Yes, done like suggested! :)

bgianfo reviewed (changes requested):
A few cases of StringView usage that need to be revisited

kennethmyhra commented:
I moved the `chmod` related commits out of this PR and will work on that in a separate branch applying Brian's suggestions.

creator1creeper1 commented:
Since commits are supposed to be atomic, shouldn't the syscall wrappers be introduced before their usage? For example `Core::System::chown()` should be introduced before using it in `chgrp`.

kennethmyhra commented:
@creator1creeper1 Thanks, fixed

creator1creeper1 commented:
This is just a minor nitpick, but you could use `:^)` consistently in the commit messages.

bgianfo commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -287,4 +287,17 @@ ErrorOr<struct passwd*> getpwnam(StringView name)
+            return gr;
+    }
+    if (errno == 0)
+        return nullptr;
```
I'm honestly not sure how this should work? The caller interprets null as error... And the normal pattern is that try returns an unwrapped value that is know to be valid. So returning a nullptr here instead of an error is kind is confusing. But... It's also how the real API works. Maybe others have  more clarity on what we should do here?

kennethmyhra replied:
That's more or less how I reasoned about this as well. Didn't feel good, but unsure how intrusive it would become if it returns an error.

kennethmyhra replied:
Would changing `getpwnam` and `getgrnam` to `Error::from_string_literal("getpwnam: Unknown username"sv);` if it doesn't find the username or group name.

```cpp
ErrorOr<struct passwd*> getpwnam(StringView name)
{
    ::setpwent();
    while (auto* pw = ::getpwent()) {
        if (pw->pw_name == name)
            return pw;
    }
    if (errno == 0)
        Error::from_string_literal("getpwnam: Unknown username"sv);
    else
        return Error::from_syscall("getpwnam"sv, -errno);
}

ErrorOr<struct group*> getgrnam(StringView name)
{
    ::setgrent();
    while (auto* gr = ::getgrent()) {
        if (gr->gr_name == name)
            return gr;
    }
    if (errno == 0)
        Error::from_string_literal("getgrnam: Unknown group name"sv);
    else
        return Error::from_syscall("getgrnam"sv, -errno);
}
```

bgianfo reviewed (approved):
Looks good, one small question...

creator1creeper1 commented:
I had left these comments earlier, but I didn't realize I had to finalize the review.

bgianfo commented on Userland/Utilities/cut.cpp:
```diff
@@ -169,19 +169,19 @@ int main(int argc, char** argv)
     if (selected_options_count == 0) {
         warnln("cut: you must specify a list of bytes, or fields");
-        args_parser.print_usage(stderr, argv[0]);
+        args_parser.print_usage(stderr, arguments.strings[0].characters_without_null_termination());
```
I think this is good feedback, but given the large scope of the PR, it would probably be better as a follow up change after this has landed?

kennethmyhra replied:
Would be nice to follow up this in a separate PR. Starting to get cross-eyed from rebasing this branch :)

kennethmyhra commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -300,4 +300,16 @@ ErrorOr<struct group*> getgrnam(StringView name)
         return Error::from_syscall("getgrnam"sv, -errno);
 }
 
+ErrorOr<void> clock_settime(clockid_t clock_id, struct timespec* ts)
```
For `clock_settime` I am just following the code pattern for similar functions above like: [stat](https://github.com/SerenityOS/serenity/blob/8595fd585df7e00447da5792e0f96987bf3d8f21/Userland/Libraries/LibCore/System.cpp#L166) and [lstat](https://github.com/SerenityOS/serenity/blob/8595fd585df7e00447da5792e0f96987bf3d8f21/Userland/Libraries/LibCore/System.cpp#L184)

The build step for lagom fails if I try to change the method to this:
```cpp
ErrorOr<void> clock_settime(clockid_t clock_id, struct timespec* ts)
{
    int rc = syscall(SC_clock_settime, clock_id, ts);
    HANDLE_SYSCALL_RETURN_VALUE("clocksettime"sv, rc, {});
}
```
```shell
[0/5] Performing build step for 'lagom'
[0/2] Re-checking globbed directories...
[6/18] Linking CXX executable Tools/CodeGenerators/LibWeb/Generate_CSS_ValueID_h
FAILED: Tools/CodeGenerators/LibWeb/Generate_CSS_ValueID_h 
: && /usr/bin/g++     Tools/CodeGenerators/LibWeb/CMakeFiles/Generate_CSS_ValueID_h.dir/Generate_CSS_ValueID_h.cpp.o  -o Tools/CodeGenerators/LibWeb/Generate_CSS_ValueID_h  -Wl,-rpath,/home/kennethmyhra/src/serenity/Build/lagom:  liblagom-core.so.0.0.0  -lpthread  -lcrypt && :
/usr/bin/ld: liblagom-core.so.0.0.0: undefined reference to `syscall2'
collect2: error: ld returned 1 exit status
[7/18] Linking CXX executable Tools/CodeGenerators/LibWeb/Generate_CSS_ValueID_cpp

```

creator1creeper1 replied:
Sure, then we should follow the established pattern.

kennethmyhra commented on Userland/Utilities/dd.cpp:
```diff
@@ -35,11 +36,9 @@ enum Status {
 };
 
-static String split_at_equals(const char* argument)
+static String split_at_equals(String argument)
```
Updated as suggested

kennethmyhra commented:
I have addressed the feedback from @bgianfo and @creator1creeper1. Also updated commit messages to consistently use :^) per recommendation. :^)

The current build fails seems unrelated to this PR. I'll follow up again when CI has completed

Build output from failing check:
```
Err:59 http://ppa.launchpad.net/canonical-server/server-backports/ubuntu focal/main amd64 libslirp0 amd64 4.4.0-1ubuntu0.21.10.1~backport20.04-202111041414~ubuntu20.04.1
  Cannot initiate the connection to ppa.launchpad.net:80 (2001:67c:1560:8008::19). - connect (101: Network is unreachable) Could not connect to ppa.launchpad.net:80 (91.189.95.85), connection timed out [IP: 91.189.95.85 80]
```

awesomekling commented:
From #code-review on Discord:

> I think we can allow ourselves to return an `Error` for the "not found" case, since that makes the API a bit more idiomatic
> then we can also return `ErrorOr<struct passwd>` instead of `ErrorOr<struct passwd*>`

awesomekling commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -308,4 +308,17 @@ ErrorOr<void> chown(StringView pathname, uid_t uid, gid_t gid)
+        if (pw->pw_name == name)
+            return *pw;
+    }
+    if (errno == 0)
```
Hmm, I'm not sure we can rely on `errno` being zero when there's no error.
My understanding is that POSIX functions are only expected to update `errno` when also returning a value that indicates an error occurred.

kennethmyhra replied:
Looking at the man page for `setpwent()` and `getpwent()`: 
> If an error occurs, errno is set appropriately.  If one wants to check errno after the  call,      it should be set to zero before the call.

Amending the function to set `errno` to zero in the beginning of the call and then checking at the end of the call seems appropriate from my understanding.

kennethmyhra replied:
Since I do not see us set `errno = 0` any other places at the beginning of calls I for now just updated it to check for `errno != 0` after `setpwent()` and then return `Error::from_syscall` if true.

kennethmyhra replied:
Scracth my comments above.

I think this should resolve errors from `setpwent` and `getpwent` in `getpwnam()` by checking `errno` after each call. Same type of implementation also in `getgrnam` https://github.com/SerenityOS/serenity/pull/11101/commits/c4d2bd2fc63bb3172b1e332139c2ed1d780a6244

bgianfo commented:
LGTM!
